File: clients/ruby/lib/pizzasql.rb
require 'net/http'
require 'uri'
require 'json'

module PizzaSQL
  class Client
    attr_reader :base_url, :db_name, :api_key

    # Creates a new PizzaSQL client connection
    #
    # @param uri [String] Database URI (e.g., 'http://localhost:8080/mydb')
    # @param api_key [String, nil] Optional API key for authentication
    # @return [Client] A new client instance
    def initialize(uri, api_key = nil)
      parsed_uri = URI.parse(uri)

      # Extract database name from path
      path = parsed_uri.path.strip.delete_prefix('/')
      raise ArgumentError, 'Database name not found in URI path' if path.empty?

      # Split path to get database name (last segment)
      path_parts = path.split('/')
      @db_name = path_parts.last

      # Reconstruct base URL without the database path
      @base_url = "#{parsed_uri.scheme}://#{parsed_uri.host}:#{parsed_uri.port}"
      @api_key = api_key
    end

    # Executes a SQL query and returns the results
    #
    # @param query [String] SQL query string
    # @return [Array<Hash>] Array of rows (each row is a hash)
    # @raise [RuntimeError] if the query fails
    def sql(query)
      uri = URI.parse("#{@base_url}/#{@db_name}/query")

      request = Net::HTTP::Post.new(uri)
      request['Content-Type'] = 'application/json'
      request['Authorization'] = "Bearer #{@api_key}" if @api_key
      request.body = { query: query }.to_json

      response = Net::HTTP.start(uri.hostname, uri.port, use_ssl: uri.scheme == 'https') do |http|
        http.request(request)
      end

      unless response.is_a?(Net::HTTPSuccess)
        raise "Request failed with status #{response.code}: #{response.body}"
      end

      result = JSON.parse(response.body)
      result['rows'] || []
    end

    # Exports database or table data
    #
    # @param table [String, nil] Table name (nil for entire database)
    # @param format [String] Export format: 'sql' or 'csv'
    # @return [String] Exported data
    # @raise [RuntimeError] if the export fails
    def export(table: nil, format: 'sql')
      params = {}
      params['table'] = table if table
      params['format'] = format if format

      query_string = URI.encode_www_form(params)
      uri = URI.parse("#{@base_url}/#{@db_name}/export?#{query_string}")

      request = Net::HTTP::Get.new(uri)
      request['Authorization'] = "Bearer #{@api_key}" if @api_key

      response = Net::HTTP.start(uri.hostname, uri.port, use_ssl: uri.scheme == 'https') do |http|
        http.request(request)
      end

      unless response.is_a?(Net::HTTPSuccess)
        raise "Export failed with status #{response.code}: #{response.body}"
      end

      response.body
    end

    # Imports data into the database
    #
    # @param data [String] Data to import
    # @param format [String] Import format: 'sql' or 'csv'
    # @param create_table [Boolean] Create table if it doesn't exist (CSV only)
    # @return [void]
    # @raise [RuntimeError] if the import fails
    def import(data, format: 'sql', create_table: false)
      params = {}
      params['format'] = format if format
      params['create_table'] = 'true' if create_table

      query_string = URI.encode_www_form(params)
      uri = URI.parse("#{@base_url}/#{@db_name}/import?#{query_string}")

      request = Net::HTTP::Post.new(uri)
      request['Content-Type'] = 'application/octet-stream'
      request['Authorization'] = "Bearer #{@api_key}" if @api_key
      request.body = data

      response = Net::HTTP.start(uri.hostname, uri.port, use_ssl: uri.scheme == 'https') do |http|
        http.request(request)
      end

      unless response.is_a?(Net::HTTPSuccess)
        raise "Import failed with status #{response.code}: #{response.body}"
      end

      nil
    end
  end

  # Module-level convenience method to create a new client connection
  #
  # @param uri [String] Database URI
  # @param api_key [String, nil] Optional API key
  # @return [Client] A new client instance
  def self.connect(uri, api_key = nil)
    Client.new(uri, api_key)
  end
end

File: clients/go/pizzasql.go
package pizzasql

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
)

// Client represents a connection to a PizzaSQL database
type Client struct {
	baseURL string
	dbName  string
	apiKey  string
	client  *http.Client
}

// Row represents a single row in the result set
type Row map[string]interface{}

// QueryResult represents the result of a SQL query
type QueryResult struct {
	Rows []Row `json:"rows"`
}

// Connect creates a new PizzaSQL client connection
// URI format: http://host:port/dbname or https://pizzabase.cloud/my_org/sql/my_db:32131
func Connect(uri string, apiKey string) (*Client, error) {
	parsedURL, err := url.Parse(uri)
	if err != nil {
		return nil, fmt.Errorf("invalid URI: %w", err)
	}

	// Extract database name from path
	path := strings.Trim(parsedURL.Path, "/")
	if path == "" {
		return nil, fmt.Errorf("database name not found in URI path")
	}

	// Split path to get database name (last segment)
	pathParts := strings.Split(path, "/")
	dbName := pathParts[len(pathParts)-1]

	// Reconstruct base URL without the database path
	baseURL := fmt.Sprintf("%s://%s", parsedURL.Scheme, parsedURL.Host)

	return &Client{
		baseURL: baseURL,
		dbName:  dbName,
		apiKey:  apiKey,
		client:  &http.Client{},
	}, nil
}

// SQL executes a SQL query and returns the results as a slice of rows
func (c *Client) SQL(query string) ([]Row, error) {
	// Prepare request body
	body := map[string]string{"query": query}
	jsonBody, err := json.Marshal(body)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	// Create request
	url := fmt.Sprintf("%s/%s/query", c.baseURL, c.dbName)
	req, err := http.NewRequest("POST", url, bytes.NewReader(jsonBody))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	// Set headers
	req.Header.Set("Content-Type", "application/json")
	if c.apiKey != "" {
		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.apiKey))
	}

	// Execute request
	resp, err := c.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	// Read response body
	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	// Check status code
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("request failed with status %d: %s", resp.StatusCode, string(respBody))
	}

	// Parse response
	var result QueryResult
	if err := json.Unmarshal(respBody, &result); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return result.Rows, nil
}

// Export exports a database or table to SQL or CSV format
func (c *Client) Export(table string, format string) ([]byte, error) {
	params := url.Values{}
	if table != "" {
		params.Set("table", table)
	}
	if format != "" {
		params.Set("format", format)
	}

	url := fmt.Sprintf("%s/%s/export?%s", c.baseURL, c.dbName, params.Encode())
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	if c.apiKey != "" {
		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.apiKey))
	}

	resp, err := c.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("export failed with status %d: %s", resp.StatusCode, string(data))
	}

	return data, nil
}

// Import imports data from SQL or CSV format
func (c *Client) Import(data []byte, format string, createTable bool) error {
	params := url.Values{}
	if format != "" {
		params.Set("format", format)
	}
	if createTable {
		params.Set("create_table", "true")
	}

	url := fmt.Sprintf("%s/%s/import?%s", c.baseURL, c.dbName, params.Encode())
	req, err := http.NewRequest("POST", url, bytes.NewReader(data))
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/octet-stream")
	if c.apiKey != "" {
		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.apiKey))
	}

	resp, err := c.client.Do(req)
	if err != nil {
		return fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("import failed with status %d: %s", resp.StatusCode, string(respBody))
	}

	return nil
}
